File: defense.py



import globals
import detection_diagnosis
import mitigation_strategy
import VM
import math
import logging

def initializeAdaptive():
	globals.DEBUG_LOGGER.debug("Function: initializeAdaptive - Initialize inital capapcity of ingress locations")
	total_num_vms = math.floor(globals.ISP_CAP*1.0/globals.VM_COMPUTE_CAP)
	for i in range(0, globals.INGRESS_LOC):
		globals.NUM_VMs.append(math.floor(total_num_vms*1.0//globals.INGRESS_LOC))
		queueSize = globals.NUM_NIC_VM*globals.NUM_VMs[i]*globals.BUFF_SIZE
		dequeuePkts = globals.NUM_NIC_VM*globals.NUM_VMs[i]
		vmCapacity = globals.NUM_VMs[i]*globals.VM_COMPUTE_CAP
		globals.INGRESS_CAP.append(VM.VM(vmCapacity,queueSize,dequeuePkts,vmCapacity))
		globals.DEBUG_LOGGER.debug(f"Function: initializeAdaptive - Initial capacity of ingress {i} is {vmCapacity}")

		# globals.INGRESS_CAP[i] = math.floor(total_num_vms/globals.INGRESS_LOC)




def initialize():
	if(globals.DEFENSE_TYPE == "adaptive"):
		initializeAdaptive()


def diagnose(pkt):
	detection_diagnosis.diagnoseTraffic(pkt)



def mitigate():
	if(globals.DEFENSE_TYPE == "static"):
		mitigation_strategy.staticMitigation()
	elif(globals.DEFENSE_TYPE == "dynamic"):
		mitigation_strategy.dynamicMitigation()
	elif(globals.DEFENSE_TYPE == "adaptive"):
		mitigation_strategy.adaptiveMitigation()

File: go/src/simulator/defense.go
package main

import "math"

type VM struct {
	cap                float64
	vmQueue            float64
	numOfDequeuePkts   int
	availableBuffSpace float64
}

func initializeDefense() {
	_DEBUG.Printf("Function: initializeAdaptive - Initialize inital capapcity of ingress locations")
	total_num_vms := CONFIGURATION.ISP_CAP / CONFIGURATION.VM_COMPUTE_CAP
	for i := 0; i < CONFIGURATION.INGRESS_LOC; i++ {
		NUM_VMs = append(NUM_VMs, int(math.Floor(total_num_vms*1.0/float64(CONFIGURATION.INGRESS_LOC))))
		queueSize := float64(CONFIGURATION.NUM_NIC_VM) * float64(NUM_VMs[i]) * CONFIGURATION.BUFF_SIZE
		virtual := float64(CONFIGURATION.NUM_NIC_VM) * float64(NUM_VMs[i]) * 80
		vmCapacity := float64(NUM_VMs[i]) * CONFIGURATION.VM_COMPUTE_CAP * float64(CONFIGURATION.NUM_NIC_VM)
		dequeuePkts := int(math.Ceil(vmCapacity / (PKT_LEN * 1000)))
		INGRESS_CAP = append(INGRESS_CAP, new(VM))
		INGRESS_CAP[i].cap = vmCapacity
		INGRESS_CAP[i].vmQueue = queueSize
		INGRESS_CAP[i].numOfDequeuePkts = dequeuePkts
		INGRESS_CAP[i].availableBuffSpace = queueSize

		_DEBUG.Printf("Function: initializeAdaptive - Initial capacity of ingress %d = %f", i, vmCapacity)
		_DEBUG.Printf("Function: initializeAdaptive - Packets to dequeue at ingress %d = %d", i, dequeuePkts)
		_DEBUG.Printf("Function: initializeAdaptive - Initial queue size at ingress %d = %f", i, queueSize)
		_INFO.Printf("BufferSize %f ingress %d", virtual, i)

		// # CONFIGURATION.INGRESS_CAP[i] = math.floor(total_num_vms/CONFIGURATION.INGRESS_LOC)
	}
}

func diagnose(pkt packet) {
	diagnoseTraffic(pkt)
}

func mitigate() {
	if CONFIGURATION.DEFENSE_TYPE == "static" {
		staticMitigation()
	} else if CONFIGURATION.DEFENSE_TYPE == "dynamic" {
		dynamicMitigation()
	} else if CONFIGURATION.DEFENSE_TYPE == "adaptive" {
		adaptiveMitigation()
	}
}
